File: Java/emails/src/main/java/com/edp/emails/consumer/EmailConsumer.java
package com.edp.emails.consumer;

import com.edp.emails.config.RabbitMQConfig;
import com.edp.emails.dto.EmailMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.stereotype.Service;

import java.util.Random;

@Service
public class EmailConsumer {
    private static final Logger logger = LoggerFactory.getLogger(EmailConsumer.class);
    private final Random random = new Random();

    @RabbitListener(queues = RabbitMQConfig.EMAIL_QUEUE)
    public void processEmail(EmailMessage message) {
        logger.info("Processing email: {}", message);

        try {
            Thread.sleep(random.nextInt(800, 1200));

            // simulating a failure
            if (shouldSimulateFailure()) {
                throw new RuntimeException("Simulated failure sending email");
            }

            switch (message.type()) {
                case CONFIRMATION -> sendConfirmationEmail(message);
                case WELCOME -> sendWelcomeEmail(message);
                case PASSWORD_RECOVERY -> sendPasswordRecoveryEmail(message);
            }

            logger.info("Email sent successfully: {}", message);
        }
        catch (Exception error) {
            logger.error("Error processing email: {}", error.getMessage());
            throw new RuntimeException(error);
        }
    }

    private void sendConfirmationEmail(EmailMessage message) {
        logger.info("Confirmation sending to: {}", message.recipient());
        logger.info("   Subject: {}", message.subject());
        logger.info("   Body: {}", message.body());
    }
    
    private void sendWelcomeEmail(EmailMessage message) {
        logger.info("Welcome sending to: {}", message.recipient());
        logger.info("   Subject: {}", message.subject());
        logger.info("   Body: {}", message.body());
    }
    
    private void sendPasswordRecoveryEmail(EmailMessage message) {
        logger.info("Password recovery sending to: {}", message.recipient());
        logger.info("   Subject: {}", message.subject());
        logger.info("   Body: {}", message.body());
    }
    
    private boolean shouldSimulateFailure() {
        // 30% chance of failure to test retry and DLQ
        return random.nextInt(100) < 30;
    }

    // Consumer for DLQ
    @RabbitListener(queues = RabbitMQConfig.DLQ_QUEUE)
    public void processFailedEmail(EmailMessage message) {
        logger.error("Email went to DLQ after 3 failed attempts!");
        logger.error("   ID: {}", message.id());
        logger.error("   Type: {}", message.type());
        logger.error("   Recipient: {}", message.recipient());
    }
}

File: Csharp/EmailsService/Services/EmailConsumer.cs
using EmailsService.Configuration;
using EmailsService.Models;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;
using System.Text;
using System.Text.Json;

namespace EmailsService.Services;

public class EmailConsumer : BackgroundService
{
    private readonly IConnection _connection;
    private readonly IModel _channel;
    private readonly ILogger<EmailConsumer> _logger;
    private readonly Random _random = new();

    public EmailConsumer(ILogger<EmailConsumer> logger)
    {
        _logger = logger;
        
        var factory = new ConnectionFactory
        {
            HostName = "localhost",
            Port = 5672,
            UserName = "guest",
            Password = "guest"
        };

        _connection = factory.CreateConnection();
        _channel = _connection.CreateModel();
        
        RabbitMQConfig.ConfigureRabbitMQ(_channel);
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        var emailConsumer = new EventingBasicConsumer(_channel);
        emailConsumer.Received += async (model, ea) =>
        {
            var body = ea.Body.ToArray();
            var message = RabbitMQConfig.DeserializeMessage<EmailMessage>(body);

            if (message != null)
            {
                await ProcessEmail(message, ea.DeliveryTag);
            }
        };

        _channel.BasicConsume(
            queue: RabbitMQConfig.EmailQueue,
            autoAck: false,
            consumer: emailConsumer
        );

        var dlqConsumer = new EventingBasicConsumer(_channel);
        dlqConsumer.Received += (model, ea) =>
        {
            var body = ea.Body.ToArray();
            var message = RabbitMQConfig.DeserializeMessage<EmailMessage>(body);

            if (message != null)
            {
                ProcessFailedEmail(message);
            }

            _channel.BasicAck(ea.DeliveryTag, false);
        };

        _channel.BasicConsume(
            queue: RabbitMQConfig.DlqQueue,
            autoAck: false,
            consumer: dlqConsumer
        );

        while (!stoppingToken.IsCancellationRequested)
        {
            await Task.Delay(1000, stoppingToken);
        }
    }

    private async Task ProcessEmail(EmailMessage message, ulong deliveryTag)
    {
        _logger.LogInformation("Processing email: {Message}", JsonSerializer.Serialize(message));

        try
        {
            await Task.Delay(_random.Next(800, 1200));

                // Simulando falha (30% de chance)
            if (ShouldSimulateFailure())
            {
                throw new Exception("Simulated failure sending email");
            }

            switch (message.Type)
            {
                case EmailType.Confirmation:
                    SendConfirmationEmail(message);
                    break;
                case EmailType.Welcome:
                    SendWelcomeEmail(message);
                    break;
                case EmailType.PasswordRecovery:
                    SendPasswordRecoveryEmail(message);
                    break;
            }

            _logger.LogInformation("Email sent successfully: {Message}", JsonSerializer.Serialize(message));
            _channel.BasicAck(deliveryTag, false);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing email: {Message}", ex.Message);
            
            _channel.BasicNack(deliveryTag, false, false);
        }
    }

    private void SendConfirmationEmail(EmailMessage message)
    {
        _logger.LogInformation("Confirmation sending to: {Recipient}", message.Recipient);
        _logger.LogInformation("   Subject: {Subject}", message.Subject);
        _logger.LogInformation("   Body: {Body}", message.Body);
    }

    private void SendWelcomeEmail(EmailMessage message)
    {
        _logger.LogInformation("Welcome sending to: {Recipient}", message.Recipient);
        _logger.LogInformation("   Subject: {Subject}", message.Subject);
        _logger.LogInformation("   Body: {Body}", message.Body);
    }

    private void SendPasswordRecoveryEmail(EmailMessage message)
    {
        _logger.LogInformation("Password recovery sending to: {Recipient}", message.Recipient);
        _logger.LogInformation("   Subject: {Subject}", message.Subject);
        _logger.LogInformation("   Body: {Body}", message.Body);
    }

    private bool ShouldSimulateFailure()
    {
        return _random.Next(100) < 30;
    }

    private void ProcessFailedEmail(EmailMessage message)
    {
        _logger.LogError("Email went to DLQ after 3 failed attempts!");
        _logger.LogError("   ID: {Id}", message.Id);
        _logger.LogError("   Type: {Type}", message.Type);
        _logger.LogError("   Recipient: {Recipient}", message.Recipient);
    }

    public override void Dispose()
    {
        _channel?.Close();
        _connection?.Close();
        base.Dispose();
    }
}

